include/envoy/http/original_ip_detection.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/http/codes.h"
+#include "envoy/http/header_map.h"
+#include "envoy/network/address.h"
+#include "envoy/server/factory_context.h"
+
+namespace Envoy {
+namespace Http {
+
+struct OriginalIPDetectionParams {
+  // The request headers from downstream.
+  //
+  // Note that while extensions can modify the headers, they will undergo standard Envoy
+  // sanitation after the detect() call so additions made here may be removed before
+  // filters have access to headers.
+  Http::RequestHeaderMap& request_headers;
+  // The downstream directly connected address.
+  const Network::Address::InstanceConstSharedPtr& downstream_remote_address;
```

Comment:
nit: if you end up doing another round I would just make this a reference to the instance vs. a reference to the shared pointer.

Reply:
given we do take references to InstanceConstSharedPtr all around the code base, passing